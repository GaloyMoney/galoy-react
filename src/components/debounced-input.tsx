import { ChangeEvent, useState, memo, useEffect } from "react"
import { useDebouncedCallback } from "use-debounce"

export type OnInputValueChange = (value: string) => void

type Props = {
  newValue?: string
  onChange?: OnInputValueChange
  onDebouncedChange?: OnInputValueChange
  [prop: string]: unknown
}

type InputObject = {
  value: string
  debouncedValue?: string
  typing: boolean
}

const DebouncedInputComponent = ({
  newValue,
  onChange,
  onDebouncedChange,
  ...inputProps
}: Props) => {
  const [input, setInput] = useState<InputObject>({ value: "", typing: false })

  const setDebouncedInputValue = useDebouncedCallback((debouncedValue) => {
    setInput((currInput) => ({ ...currInput, debouncedValue, typing: false }))
  }, 1500)

  useEffect(() => {
    if (newValue !== undefined) {
      setInput({ value: newValue, typing: false })
    }
  }, [newValue])

  useEffect(() => {
    if (input.typing) {
      setDebouncedInputValue(input.value)
    }
    return () => setDebouncedInputValue.cancel()
  }, [setDebouncedInputValue, input.typing, input.value])

  useEffect(() => {
    if (onDebouncedChange && input.debouncedValue !== undefined) {
      onDebouncedChange(input.debouncedValue)
    }
  }, [onDebouncedChange, input.debouncedValue])

  const handleOnChange = (event: ChangeEvent<HTMLInputElement>) => {
    const eventValue = event.target.value
    if (onChange) {
      onChange(eventValue)
    }
    setInput({ value: eventValue, typing: true })
  }

  return <input value={input.value} onChange={handleOnChange} {...inputProps} />
}

export const DebouncedInput = memo(DebouncedInputComponent)
