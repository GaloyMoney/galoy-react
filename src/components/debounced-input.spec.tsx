import { render } from "@testing-library/react"

import { DebouncedInput } from "./debounced-input"

describe("DebouncedInput", () => {
  it("renders and matches snapshot", () => {
    const { asFragment } = render(<DebouncedInput />)
    expect(asFragment()).toMatchSnapshot()
  })
})
