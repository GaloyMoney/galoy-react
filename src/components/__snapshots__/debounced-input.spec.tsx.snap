// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DebouncedInput renders and matches snapshot 1`] = `
<DocumentFragment>
  <input
    value=""
  />
</DocumentFragment>
`;
